### PR TITLE
Send invite Reminders

### PR DESCRIPTION
**Pull request by hardillb**
part of #4380

## Description

Sends reminder emails to invitees if not actioned in 2 days.

Also includes a house keeping job to remove expired invites from the database.

~~Still need to also send email to inviter to say not accepted yet.~~

~~Also needs tests, but will need to work out how to inject invites into the database with a custom `createdAt` entry in the database.~~

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by knolleary** [comment]
Looks okay from a first pass through. Want to sanity check the right emails are sent to the right parties - I note the tests just verify the right number of emails are sent, not that the right content was sent to the right recipient.

We also need to consider the scaling issue. Is the easy option just to add a 'reminderSent' column to the table?

**Comment by hardillb** [comment]
We could add a column to the Invitations table, but we'd have to have it run multiple times a day to reduce the chance of restarts meaning that we never sent the reminders.

I'll add some more to the tests to check the email contents

**Comment by hardillb** [comment]
OK, I have a version that uses a db column to record when the invite was sent.

I'm running the job with the following cron job

```
    // runs every 15mins with a random offset from the start of the hour
    schedule: `${randomInt(0, 15)}/15 * * * *`,
```

This will run at 15 min intervals, with a random offset between 0 & 15mins. This should mean that different instances run at different times, but that it not skip a day if the app is restarted. It may be a bit too many times, but it should only send reminders once, so should be a null op most of the time.

File: forge/housekeeper/tasks/inviteReminder.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+const { Op } = require('sequelize')
+
+const { randomInt } = require('../utils')
+
+module.exports = {
+    name: 'inviteReminder',
+    startup: false,
+    // runs every 15mins with a random offset from the start of the hour
+    schedule: `${randomInt(0, 15)}/15 * * * *`,
```
Comment (knolleary) [inline forge/housekeeper/tasks/inviteReminder.js:9]:
```suggestion
    // runs daily at a randomly picked time
    schedule: `${randomInt(0, 59)} 6 * * *`,
```

Not sure why this needs to run so frequently? Once a day feels sufficient. Or am I missing a scenario that requires this to run more often?

Comment (hardillb) [reply forge/housekeeper/tasks/inviteReminder.js:9]: OK, so pinning it to a given hour means that the window to not run updates is a restart inside that hour.

Making it 6am is probably early enough that we are unlikely to do a code push/restart in that hour.

**Comment by knolleary** [comment]
Testing locally, I see the emails get sent, but I consistently get this error in the logs:
```
[2024-12-05T11:39:00.066Z] TRACE: Completed task 'inviteReminder'
[2024-12-05T11:39:00.503Z] WARN: Failed to send email: Error: Message failed: 450 SQLITE_ERROR: cannot start a transaction within a transaction
```

The Invitation's `reminderSentAt` column has been updated, so not sure what DB interaction is triggering this.

**Comment by knolleary** [comment]
`WARN: Failed to send email: Error: Message failed: 450 SQLITE_ERROR: cannot start a transaction within a transaction`

I think this is a complete red herring. The error is the response being returned by my local test SMTP server. From our logs, two mails were sent. Looking at the inbox, only one arrived.

**Comment by knolleary** [comment]
Yup - upstream issue with my choice of SMTP tools - https://github.com/nodemailer/nodemailer-app/issues/3

**Review by knolleary — APPROVED** [review]

**Pull request merged**